### PR TITLE
Non display channel group data

### PR DESCRIPTION
**Pull request by Paul-Licameli**
Resolves: #2790 
Resolves: #2804

Move many things into ChannelGropuAttachments, so they are not duplicated between channels.

This simplifies code that formerly had to maintain consistency of channels.  There are fewer uses of TrackList::Channels
and fewer iterations over all tracks when iteration over leaders only is enough.

Remove some other calls to TrackList::Channels() that were needed only to count the channels.

Also fixes an issue related to correctness of Mute and Solo settings when copying and pasting or duplicating MIDI or
audio tracks.

Also a new small library lib-playable-track to lift the mute and solo logic out of lib-track.

To test:
- [ ] Macro commands "Set Clip", "Set Envelope", "Set Track Status", "Set Track Audio", "Set Track Visuals", "Set Track"
- [x] copy and paste preserves mono or stereo track name, pan, gain, mute, solo
- [x] Changing mono or stereo track name, pan, gain, mute, solo
- [x] Splitting stereo track to mono replicates name, pan, gain, mute, solo*
- [x] "Solo Button" setting in TracksBehaviors works as before
- [x] Importing mono or stereo into a project with a Solo track still makes all new tracks mute
- [x] Shift+click on Mute and Solo buttons works as before
- [x] Pan, gain, mute, solo in the Mixer Board window
- [x] Pan, gain, mute, solo affect playback and mix-and-render as expected
- [x] Changing spectrogram scale type with right-click in either channel's vertical ruler affects both channels
- [x] Changing spectrogram bounds with zooming commands, or scroll wheel, or click and drag (turn on Advanced Vertical Zooming in Tracks Behaviors preferences), in either channel's ruler affects both
- [x] Similarly for changes of Waveform scale type or bounds (including changes of bottom of dB range with scroll wheel)
- [x] For split wave/spectrum view, changing which is on top (the hand cursor), or proportion of the split view (dragging the separator), in either channel, changes both
- [x] Changing among Waveform, Spectrum, Multi view using drop-down menu, mono and stereo
- [x] Track pan, gain, solo and mute states are preserved upon save, close & re-open
- [x] Mixing of tracks create new tracks with expected pan, gain, solo and mute states

* Possible non-uniqueness of the solo track may be considered a bug to fix, but not a regression caused by this pull request

- [x] I signed [CLA](https://www.audacityteam.org/cla/)
- [x] The title of the pull request describes an issue it addresses
- [x] If changes are extensive, then there is a sequence of easily reviewable commits
- [x] Each commit's message describes its purpose and effects
- [x] There are no behavior changes unnecessary for the stated purpose of the PR

Recommended:
- [x] Each commit compiles and runs on my machine without known undesirable changes of behavior

### CODE REVIEW

**Comment by saintmatthieu** [comment]
[] There are track selection commands whose implementation were modified by `264ea9f`. I couldn't find them anywhere in the UI so don't know how to test this but I suppose it is testable.

**Comment by saintmatthieu** [comment]
@Paul-Licameli in the test list you mention
> - [ ] Splitting stereo track to mono replicates name, pan, gain, mute, solo*
Why specifically to mono? Selectness should also be replicated.

**Comment by saintmatthieu** [comment]
Thanks for the `PlayableTrack.h` clean-up.

**Comment by saintmatthieu** [comment]
- [ ] I built `a95e16c` and crashed it by opening a one-track Audacity 3.2 project.

image

**Review by saintmatthieu — CHANGES_REQUESTED** [review]
I stopped before the "Don't loop" commits, but there's already comments I'd like you to look at.

**Comment by saintmatthieu** [comment]
Set status to "request changes" essentially for the bug. Not sure if the rest of my comments should result in changes yet, awaiting your replies for that.

**Comment by Paul-Licameli** [comment]
> @Paul-Licameli in the test list you mention
> 
> > Splitting stereo track to mono replicates name, pan, gain, mute, solo*
> 
> * [ ]  Why specifically to mono? Selectness should also be replicated.

I updated the top comment to mention selected state too.

"Why specifically to mono?"  I don't understand, what other kind of splitting is there?

File: libraries/lib-playable-track/PlayableTrack.h [hunk 1]
```diff
@@ -76,4 +77,13 @@ class PLAYABLE_TRACK_API PlayableTrack /* not final */ : public AudioTrack
...
 
 ENUMERATE_TRACK_TYPE(PlayableTrack);
+
+enum SoloBehavior {
+   SoloBehaviorSimple,
+   SoloBehaviorMulti,
+   SoloBehaviorNone,
+};
+
+extern PLAYABLE_TRACK_API EnumSetting<SoloBehavior> TracksBehaviorsSolo;
```
[the remark this answers is not in the document]
Comment (Paul-Licameli) [reply libraries/lib-playable-track/PlayableTrack.h:75]: See this line in TracksBehaviorsPrefs.cpp:
```
         S.TieChoice( XXO("Solo &Button:"), TracksBehaviorsSolo);
```
The second argument is by reference.  That code constructs a dialog that may modify the setting.

**Comment by Paul-Licameli** [comment]
> Set status to "request changes" essentially for the crash. There's not much I can test until it's fixed. Not sure if the rest of my comments should result in changes yet, awaiting your replies for that.

We know the fix for the crash now.  I pushed it, and also some of your less important suggested changes.

**Comment by saintmatthieu** [comment]
> "Why specifically to mono?" I don't understand, what other kind of splitting is there?

I meant "Split Stereo Track" as opposed to "Split Stereo to Mono", but actually the former looks just like the latter, with hard left-right panning. Please ignore.

**Comment by saintmatthieu** [comment]
There remains a `WaveTrack::Attachments::RegisteredFactory`, namely in `WaveTrackLocation`. Just glancing at it looks like it has to do with cut lines and merge points. Still not sure what that is and if it makes sense to still have this data on a per-channel basis or if it also should be grouped in `ChannelGroupData`.

**Comment by saintmatthieu** [comment]
There remain calls to
* `TrackList::Channels(track).size()` in Normalize.cpp, line 177,
* `TrackList::Channels(pTrack).size()` in WaveTrackControls, line 1031.

**Comment by saintmatthieu** [comment]
Sorry, I thought "Close with comment" would finalize my review, not close the PR 😅

**Review by saintmatthieu — APPROVED** [review]
Looks like we're on the right track :D Please check my comment about remaining calls to `Channels(...).size()` before merging.

**Comment by Paul-Licameli** [comment]
> There remains a `WaveTrack::Attachments::RegisteredFactory`, namely in `WaveTrackLocation`. Just glancing at it looks like it has to do with cut lines and merge points. Still not sure what that is and if it makes sense to still have this data on a per-channel basis or if it also should be grouped in `ChannelGroupData`.

That is a different attachment place, unlike the group data.

Cut line locations are still inherently per-channel rather than per-track, until I really do the reorganization of clips to contain multiple sequences.  Until then, it is possible that cut lines in opposite channels are at different times.

The time to reorganize this is not yet now.

**Comment by Paul-Licameli** [comment]
> There remain calls to
> 
> * `TrackList::Channels(track).size()` in Normalize.cpp, line 177,
> * `TrackList::Channels(pTrack).size()` in WaveTrackControls, line 1031.

Done.

**Comment by Paul-Licameli** [comment]
Interactively rebased onto latest master, squashing some fix-ups.  Not expecting anything to behave differently in the next build.

**Review by saintmatthieu — APPROVED** [review]

**Pull request merged**